### PR TITLE
Prevent crash when trying to resume audiobook with no chapters

### PR DESCRIPTION
**Pull request by dimitry-ishenko**
## Description

Transform "while" loop into "for" loop to prevent null pointer dereference.

## Motivation and Context

This change prevents Kodi from crashing when trying to resume an audiobook file with no chapters. Please see issue #15203 for details.

## How Has This Been Tested?

Tried playing the same file that earlier caused the crash.

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by DaVukovic** [comment]
I applied this patch onto current master, but Kodi still crashes with an mka file with no chapters. It crashes for me while stopping the playback or selecting any other file to play. 

http://paste.ubuntu.com/p/FcpVpMgbPC/

**Comment by DaVukovic** [comment]
That's an example file I use here. A simple mp3 converted to a mka.

[01 - Aerosmith - Intro.mka.zip](https://github.com/xbmc/xbmc/files/2735494/01.-.Aerosmith.-.Intro.mka.zip)

File: xbmc/music/windows/GUIWindowMusicBase.cpp
Comment (pkerling) [inline xbmc/music/windows/GUIWindowMusicBase.cpp]: I suggest using `std::find_if` here

btw, the missing `break` would also be solved by that :-)

**Comment by dimitry-ishenko** [comment]
> I applied this patch onto current master, but Kodi still crashes with an mka file with no chapters. It crashes for me while stopping the playback or selecting any other file to play.
> 
> http://paste.ubuntu.com/p/FcpVpMgbPC/

@DaVukovic that's a different error caused here:
https://github.com/xbmc/xbmc/blob/adb55e29fdd5e965756c876861adbd615d84cf25/xbmc/music/MusicDatabase.cpp#L10709-L10717

Your file is missing album and artist info and the above code doesn't check for that, again causing segfault on line 10713.

I should probably add it to my patch though.

File: xbmc/music/windows/GUIWindowMusicBase.cpp
Comment (dimitry-ishenko) [reply xbmc/music/windows/GUIWindowMusicBase.cpp]: @pkerling I like your idea and redid it with `std::find_if`. Please see below.

**Comment by dimitry-ishenko** [comment]
I apologize for all the whitespace cleaning -- my editor is set up to automatically do it. Let me know if that's a problem and I can redo.

File: xbmc/music/MusicDatabase.cpp
Comment (pkerling) [inline xbmc/music/MusicDatabase.cpp]:
```suggestion
    artists.empty() ? "" : artists.front().data(),
```

`empty` is more descriptive than a ternary on `size` (not tested though)

File: xbmc/music/MusicDatabase.cpp
Comment (pkerling) [inline xbmc/music/MusicDatabase.cpp]: Why did you change from `c_str()` to `data()`?

File: xbmc/music/windows/GUIWindowMusicBase.cpp
Comment (pkerling) [inline xbmc/music/windows/GUIWindowMusicBase.cpp]:
```suggestion
      auto itemIt = std::find_if(
```

code style

File: xbmc/music/windows/GUIWindowMusicBase.cpp
Comment (pkerling) [inline xbmc/music/windows/GUIWindowMusicBase.cpp]:
```suggestion
      if(itemIt != m_vecItems->end())
```

File: xbmc/music/windows/GUIWindowMusicBase.cpp
Comment (pkerling) [inline xbmc/music/windows/GUIWindowMusicBase.cpp]:
```suggestion
        CFileItem resItem(**itemIt);
```

File: xbmc/music/windows/GUIWindowMusicBase.cpp
Comment (pkerling) [inline xbmc/music/windows/GUIWindowMusicBase.cpp]:
```suggestion
        resItem.SetProperty("StartPercent", (static_cast<double> (bookmark-resItem.m_lStartOffset) / (resItem.m_lEndOffset - resItem.m_lStartOffset) * 100);
```

code style

**Review by yol — CHANGES_REQUESTED** [review]
Whitespace fixes are fine, but need to be put into a separate commit to reduce clutter for review

File: xbmc/music/MusicDatabase.cpp
Comment (dimitry-ishenko) [reply xbmc/music/MusicDatabase.cpp]: OK

File: xbmc/music/MusicDatabase.cpp
Comment (dimitry-ishenko) [reply xbmc/music/MusicDatabase.cpp]: Starting with c++11 they are equivalent. `data()` seems to be getting more love recently. Eg, `std::vector` now has `data()`.

File: xbmc/music/windows/GUIWindowMusicBase.cpp
Comment (dimitry-ishenko) [reply xbmc/music/windows/GUIWindowMusicBase.cpp]: Oh right, we do camel here.

File: xbmc/music/windows/GUIWindowMusicBase.cpp
Comment (dimitry-ishenko) [reply xbmc/music/windows/GUIWindowMusicBase.cpp]: That's @notspiff 's fault :stuck_out_tongue: but I will fix it

Comment (dimitry-ishenko) [reply xbmc/music/windows/GUIWindowMusicBase.cpp]: Agreed. Just realized the cast is not even necessary if we replace `100` with `100.0`, which will trigger implicit promotion to double.

**Comment by DaveTBlake** [comment]
>I apologize for all the whitespace cleaning -- my editor is set up to automatically do it. Let me know if that's a problem and I can redo.

Sorry I disagree with @pkerling, I really don't even want a separate commit tiding up the whitespace! Keep this PR to just audiobook changes/fixes please. MusicDatabase.cpp is huge, and I work on it lots, for clarity and my sanity any white space tidy-up is best as a separate PR. Meanwhile I will hold up my hand for probably being the source of much of that messy whitespace.

**Comment by dimitry-ishenko** [comment]
Here ya go. Hopefully everyone is happy now :smiley:

**Comment by pkerling** [comment]
> Sorry I disagree with @pkerling, I really don't even want a separate commit tiding up the whitespace! Keep this PR to just audiobook changes/fixes please.

I don't care if it's a separate PR or not, it should just definitely not be in the same commit.

File: xbmc/music/windows/GUIWindowMusicBase.cpp
Comment (pkerling) [inline xbmc/music/windows/GUIWindowMusicBase.cpp]:
```suggestion
      if (itemIt != m_vecItems->end())
```

sorry, slipped through - everything else is good to go :+1:

**Comment by pkerling** [comment]
@MartijnKaijser Crash fix, so IMO good for v18

**Comment by DaveTBlake** [comment]
Lets get jenkins green before merge
jenkins build this please

File: xbmc/music/windows/GUIWindowMusicBase.cpp
Comment (Rechi) [inline xbmc/music/windows/GUIWindowMusicBase.cpp]: As the `std::find_if` call never should change any items, it would be better to make that explicit by the following change
```suggestion
        m_vecItems->cbegin(),
```

File: xbmc/music/windows/GUIWindowMusicBase.cpp
Comment (Rechi) [inline xbmc/music/windows/GUIWindowMusicBase.cpp]:
```suggestion
        m_vecItems->cend(),
```

File: xbmc/music/windows/GUIWindowMusicBase.cpp
Comment (Rechi) [inline xbmc/music/windows/GUIWindowMusicBase.cpp]:
```suggestion
      if (itemIt != m_vecItems->cend())
```

**Comment by dimitry-ishenko** [comment]
Here you go

**Review by yol — APPROVED** [review]

**Comment by pkerling** [comment]
jenkins build this please

**Comment by pkerling** [comment]
@dimitry-ishenko Thanks for sticking through! We tend to overcompensate for our dated codebase by being quite picky in new PRs, but I hope you get the idea that we do this to have better maintainability in the future. Looking forward to seeing further contributions from you :-)

**Pull request merged**
